Initial Docs (#64)
janus-idp/operator: docs/developer.md

# DEVELOPER GUIDE --- WIP

## Contributing
// TODO(user): Add detailed information on how you would like others to contribute to this project

### How it works
This project aims to follow the Kubernetes [Operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/).

It uses [Controllers](https://kubernetes.io/docs/concepts/architecture/controller/)
which provides a reconcile function responsible for synchronizing resources until the desired state is reached on the cluster.

### Running on the cluster
1. Install Instances of Custom Resources:
```sh
kubectl apply -f config/samples/
```
2. Build and push your image to the location specified by `IMG`:
```sh
make docker-build docker-push IMG=<some-registry>/backstage-operator:tag
```
3. Deploy the controller to the cluster with the image specified by `IMG`:
```sh
make deploy IMG=<some-registry>/backstage-operator:tag
```
### Uninstall CRDs
To delete the CRDs from the cluster:
```sh
make uninstall
```
### Undeploy controller
UnDeploy the controller from the cluster:
```sh
make undeploy
```

> ## Build and Deploy with OLM

Reviewer: ```suggestion
### Build and Deploy with OLM
```

This should be at the same level as the rest of the topics, right? Otherwise, the table of contents seems strange to me:

![image](https://github.com/janus-idp/operator/assets/593208/2f140a33-abf7-4df3-b7d4-0ae00bcc9f58)

Author: I think so, it was generated in fact :)

# DEVELOPER GUIDE --- WIP

## Contributing
// TODO(user): Add detailed information on how you would like others to contribute to this project

### How it works
This project aims to follow the Kubernetes [Operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/).

It uses [Controllers](https://kubernetes.io/docs/concepts/architecture/controller/)
which provides a reconcile function responsible for synchronizing resources until the desired state is reached on the cluster.

## Local development

### Prerequisites

> You need ready to use **kubectl** and **minikube** [tools](https://kubernetes.io/docs/tasks/tools/) as a prerequisite.

Reviewer 2: We should strive to make this doc more consistent with the format we use downstream, eg., in https://access.redhat.com/documentation/en-us/red_hat_openshift_dev_spaces/3.10/html-single/administration_guide/index#installing-devspaces-in-a-restricted-environment

```suggestion
* **kubectl**. See [Instaling kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl).
* **minikube**. See [Instaling minkube](https://kubernetes.io/docs/tasks/tools/#minikube).
```

> Get your copy of Backstage Operator from GitHub:

Reviewer 2: ```suggestion
* A copy of the Backstage Operator sources:
```

```sh
git clone https://github.com/janus-idp/operator
```

### Local Tests

> Including unit testing (starting from v 0.02) and Kubernetes intergation testing ([envtest](https://book.kubebuilder.io/reference/envtest.html))

Reviewer 2: ```suggestion
To run both unit tests (@since 0.0.2) and Kubernetes integration tests ([envtest](https://book.kubebuilder.io/reference/envtest.html)): 
```

```sh
make test
```

> will run both of them  

Reviewer 2: ```suggestion
```

### Test on the local cluster

> You’ll need a Kubernetes cluster to run against. You can use [Minikube](https://minikube.sigs.k8s.io/docs/) or [KIND](https://sigs.k8s.io/kind) to get a local cluster for testing, or run against a remote cluster.

Reviewer 2: Should we link to https://kubernetes.io/docs/tasks/tools/ instead, since that's what you have above for the prereqs?
